internal/ethereum/confirmation_reconciler.go

```
@@ -52,71 +59,238 @@ func (bl *blockListener) reconcileConfirmationsForTransaction(ctx context.Contex
+	// Secondary objective is to report whether any fork was detected (and corrected) during this analysis
+	// Initialize the result with existing confirmations
+	reconcileResult := &ffcapi.ConfirmationUpdateResult{
+		TargetConfirmationCount: targetConfirmationCount,
+	}
+
+	// before we get into the main algorithm, a couple of special cases to optimize for that can save us some work
+	reconcileResult, err := bl.handleSpecialCases(ctx, existingConfirmations, txBlockInfo, targetConfirmationCount, reconcileResult)
+	if reconcileResult.Confirmed || err != nil {
+		return reconcileResult, err
+	}
+
+	// We start by constructing 2 lists of blocks:
+	// - The `earlyList`.  This is the set of earliest blocks we are interested in.  At the least, it starts with the transaction block
+	//    and may also contain some number of existing confirmations i.e. the output from previous call to this function
+	//    Other than the `transactionBlock`, we don't yet know whether any of the early list is still correct as per the current state of the canonical chain.
+	//    The chain may have been re-organized since we discovered the blocks in that list.
+	// - The `lateList`. This is the most recent set of blocks that we are interesting in and we believe are accurate for the current state of the chain
+
+	earlyList := createEarlyList(ctx, existingConfirmations, txBlockInfo, reconcileResult)
+	lateList, err := createLateList(ctx, txBlockInfo, targetConfirmationCount, reconcileResult, bl)
+	if err != nil {
+		return nil, err
+	}
+
+	// These 2 lists may overlap so we splice them together which will remove any overlapping blocks
+	splicedList, detectedFork := newSplice(earlyList, lateList)
+	if detectedFork {
+		reconcileResult.NewFork = true
+	}
+	for {
+		// now loop until we can form a contiguous linked list from the spliced list
+
+		if !splicedList.containsTransactionBlock() {
```

I'm not sure this can be covered by any test due to the behaviour in `createEarlyList`

---

i see this is because `removeBrokenLink` can remove the tx block as well

---

This should throw an error to indicate reconciliation failure